accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,523 @@
+
+Trusted Publishers will have the biggest impact on the ecosystem when it is supported on NuGet.org. This only needs the
+first branch of auth (API-key) so considering read-only (package download/restore) scenarios is not needed now.
+
+### Entra ID authentication
+
+We could opt to enable Entra ID service principal-based authentication. Then, users could leverage existing Entra ID
+support for workload identity federation
+([docs](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation)) which uses Federated Identity
+Credentials ([docs](https://learn.microsoft.com/en-us/graph/api/resources/federatedidentitycredentials-overview)). This
+is similar to [VS Marketplace](https://marketplace.visualstudio.com/) and how it could provide extension publishing from
+CI/CD workflows. In the end, this approach could support GitHub Actions via
+[`azure/login`](https://github.com/azure/login) or any compute environment you want (such as an Azure VM with a managed
+identity), as long as you can authenticate as an Entra ID service principal. The drawback of this is that NuGet.org
+would have no control over the workload identity tokens which are accepted and would be opening up a lot more
+authentication scenarios than we really need at this point. This would also require us rationalizing how an Entra ID
+service principal relates to a NuGet.org user or organization profile, which is a lot more work! Maybe it's something we
+need in the distant future, but not now.
+
+## Prior Art
+
+- This feature has been prototyped and demoed to the NuGet team a couple of times, as far back as February of 2022. More
+  recently, a prototype was made using credential providers and one-legged auth:
+  - [joelverhagen/token-login-sample](https://github.com/joelverhagen/token-login-sample) - sample one-legged auth
+    GitHub Actions workflow
+  - [joelverhagen/token-login](https://github.com/joelverhagen/token-login) - sample one-legged auth step (used by the
+    above)
+  - [NuGet/NuGetGallery@jver-oidc](https://github.com/NuGet/NuGetGallery/tree/jver-oidc) - branch on NuGetGallery to
+    supported one-legged auth
+- [Trusted Publishers for All Package
+Repositories](https://repos.openssf.org/trusted-publishers-for-all-package-repositories), by Seth Larson
+- Trusted Publishers has been implemented by:
+  - Python Package Index, a.k.a. PyPI ([blog](https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/),
+    [docs](https://docs.pypi.org/trusted-publishers/))
+  - RubyGems ([blog](https://blog.rubygems.org/2023/12/14/trusted-publishing.html),
+    [docs](https://guides.rubygems.org/trusted-publishing/))
+  - pub.dev ([docs](https://dart.dev/tools/pub/automated-publishing))
+
+## Unresolved Questions
+
+Ask away!
+
+## Future Possibilities
+
+- Enable additional CI/CD systems that support OIDC tokens
+  - Azure DevOps is next on the popularity list. Their workload identity federation is very focused on Azure today
```

Comment:
+1 to AzDo based on usage and qualitative data posted in this PR thread as a comment screenshot.

Reply:
I hope we can get Azure DevOps integration. That would help many Microsoft teams who publish from Azure DevOps to NuGet.org (e.g. NuGet client team, NuGet.* packages!).